src/embedded_flash/syscall.rs

```diff
@@ -210,6 +213,117 @@ fn find_slice<'a>(
+    page_size: usize,
+    partition_locations: Vec<PartitionSlice>,
+    metadata_location: PartitionSlice,
+}
+
+impl Partition {
+    /// Provides access to the other partition if available.
+    ///
+    /// # Errors
+    ///
+    /// Returns `CustomError` if any of the following conditions do not hold:
+    /// - The page size is a power of two.
+    /// - The storage slices are page-aligned.
+    pub fn new() -> StorageResult<Partition> {
+        let page_size = get_info(command_nr::get_info_nr::PAGE_SIZE, 0)?;
+        if !page_size.is_power_of_two() {
+            return Err(StorageError::CustomError);
+        }
+        let mut partition_locations = Vec::new();
+        let mut metadata_location = None;
+        for i in 0..memop(memop_nr::STORAGE_CNT, 0)? {
+            let storage_type = memop(memop_nr::STORAGE_TYPE, i)?;
+            match storage_type {
+                storage_type::STORE => continue,
+                storage_type::PARTITION | storage_type::METADATA => (),
+                _ => return Err(StorageError::CustomError),
+            };
+            let storage_ptr = memop(memop_nr::STORAGE_PTR, i)?;
+            let storage_len = memop(memop_nr::STORAGE_LEN, i)?;
+            if !is_aligned(page_size, storage_ptr) || !is_aligned(page_size, storage_len) {
+                return Err(StorageError::CustomError);
+            }
+            match storage_type {
+                storage_type::PARTITION => partition_locations.push(PartitionSlice {
+                    ptr: storage_ptr,
+                    len: storage_len,
+                }),
+                // We only expect one page of metadata.
+                storage_type::METADATA => {
+                    if metadata_location.is_none() {
+                        metadata_location = Some(PartitionSlice {
+                            ptr: storage_ptr,
+                            len: storage_len,
+                        });
+                    } else {
+                        return Err(StorageError::CustomError);
+                    }
+                }
+                _ => (),
+            };
+        }
+        if let Some(metadata_location) = metadata_location {
+            Ok(Partition {
+                page_size,
+                partition_locations,
+                metadata_location,
+            })
+        } else {
+            Err(StorageError::CustomError)
+        }
+    }
+
+    pub fn is_page_in_partition(&self, page_address: usize) -> bool {
+        self.partition_locations
+            .iter()
+            .any(|storage_location| storage_location.contains(page_address))
+    }
+
+    pub fn is_page_in_metadata(&self, page_address: usize) -> bool {
+        self.metadata_location.contains(page_address)
+    }
+
+    pub fn rewrite_page(&self, page_ptr: usize, value: &mut [u8]) -> StorageResult<()> {
+        let value_len = value.len();
+        if !is_aligned(self.page_size, page_ptr) || !is_aligned(self.page_size, value_len) {
```

Comment:
The caller has to check `is_page_in_partition` and `is_page_in_metadata`. Duplicating the check inside `rewrite_page` wastes some time, but prevents accidental writes to PersistentStore locations. Trying to write to any other locations is generally prevented by the MPU.
We could also decide to make the API more specific and have `rewrite_metadata` and `rewrite_partition`. We then wouldn't need to pass in the metadata address and could let callers only specify a partition offset instead of an absolute address. (And add more features, like writing a dirty bit into the metadata when writing to partitions.) Questions is if we want to handle this here or in the app.

Reply:
It's usually a good idea to avoid preconditions (in particular if they are not checked). In particular when we don't see the call-sites because we can't check that they are satisfied.

There is no need to have duplicate checks. If the function is safe, then there's no need to check before calling it, it would just return an error.

It sounds like having the API you propose would be better (and is much less low-level like `rewrite_page`):
```rust
trait UpgradeStorage {
    fn read_metadata(&self) -> StorageResult<&[u8]>;
    // error if page.len() != PAGE_SIZE
    fn write_metadata(&mut self, page: &[u8]) -> StorageResult<()>;
    // error if page_index >= num_partition_pages
    fn read_partition(&self, page_index: usize) -> StorageResult<&[u8]>;
    // error if page_index >= num_partition_pages || page.len() != PAGE_SIZE
    fn write_partition(&mut self, page_index: usize, page: &[u8]) -> StorageResult<()>;
    fn num_partition_pages(&self) -> usize;
}
```

Reply:
I implemented something similar, see the general explanation for details.